### PR TITLE
Adds option for array bounds checking for host_device_ptr on the host

### PR DESCRIPTION
**Pull request by adayton1**

### CODE REVIEW

File: src/care/host_device_ptr.h [hunk 1]
```diff
@@ -196,6 +196,22 @@ namespace care {
          return *reinterpret_cast<host_device_ptr<const T> const *> (this);
       }
 
+#if defined(CARE_ENABLE_BOUNDS_CHECKING)
+      template <class Index>
+      inline void boundsCheck(const Index i) const {
+         if (i < 0 || i >= (Index) MA::size()) {
```
Comment (liu15) [inline src/care/host_device_ptr.h:202]: what happens if the array is created as one type by accessed as another (allocated a char but accessed as an int)?

Comment (adayton1) [reply src/care/host_device_ptr.h]: Good point. Now using the size in bytes divided by the size of the type that it is when this operator is called.

**Review by liu15 — COMMENTED** [review]
Al will love this!  Any chance something like this.

Any chance something like this would ever be upstreamed into CHAI?

**Comment by adayton1** [comment]
So, there is a branch in CHAI where this was started, but it stalled because it wasn't obvious how to get this to work on the GPU.

**Comment by adayton1** [comment]
Yeah, Al was the one who asked me for this :)

**Review by dtaller — APPROVED** [review]
Looks great! Thanks! Just one question about how this is used in pick() vs operator[], besides that it is all good

File: src/care/host_device_ptr.h [hunk 2]
```diff
@@ -206,11 +222,19 @@ namespace care {
 #if !CARE_LEGACY_COMPATIBILITY_MODE
       const
 #endif
-      { return MA::operator[](i); }
+      {
+#if !defined(CARE_DEVICE_COMPILE) && defined(CARE_ENABLE_BOUNDS_CHECKING)
+         boundsCheck(i);
+#endif
+         return MA::operator[](i);
```
Comment (dtaller) [inline src/care/host_device_ptr.h:229]: Does this need to be added to the pick() function separately?

File: src/care/host_device_ptr.h [hunk 3]
```diff
@@ -206,11 +222,19 @@ namespace care {
...
+      }
 
 #if CARE_LEGACY_COMPATIBILITY_MODE
      template<typename Idx>
      inline CARE_HOST_DEVICE T& operator[](const Idx i) const {
+#if !defined(CARE_DEVICE_COMPILE) && defined(CARE_ENABLE_BOUNDS_CHECKING)
+        boundsCheck(i);
```
Comment (dtaller) [inline src/care/host_device_ptr.h:236]: Does this need to be added to the pick() function separately?

**Review by dtaller — COMMENTED** [review]
Do these need to be added to the pick() function separately, rather than just operator[]?

File: src/care/host_device_ptr.h (hunk 3 above)
Comment (adayton1) [reply src/care/host_device_ptr.h:236]: Done.

**Pull request merged**